netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -492,5 +540,33 @@ public void onPingAckRead(ChannelHandlerContext ctx, ByteBuf data) throws Http2E
         logger.warning("Received unexpected ping ack. No ping outstanding");
       }
     }
+
+    @Override
+    public void onPingRead(ChannelHandlerContext ctx, ByteBuf data) throws Http2Exception {
+      if (keepAliveManager != null) {
+        keepAliveManager.onDataReceived();
+      }
+      super.onPingRead(ctx, data);
+    }
+  }
+
+  private final class KeepAlivePinger implements KeepAliveManager.KeepAlivePinger {
+    final ChannelHandlerContext ctx;
+
+    KeepAlivePinger(ChannelHandlerContext ctx) {
+      this.ctx = ctx;
+    }
+
+    @Override
+    public void ping() {
+      encoder().writePing(ctx, false /* isAck */, KEEPALIVE_PING_BUF, ctx.newPromise());
+      ctx.flush();
+    }
+
+    @Override
+    public void onPingTimeout() {
+      ctx.channel().writeAndFlush(new ForcefulCloseCommand(
```

Comment:
This is strange, since we're already in the handler. Just call forcefulClose directly?